botocore/handlers.py

```diff
@@ -140,6 +143,28 @@ def conditionally_calculate_md5(params, **kwargs):
         calculate_md5(params, **kwargs)
 
 
+def validate_bucket_name(params, **kwargs):
+    # The rules for bucket names in the US Standard region allow bucket names
+    # to be as long as 255 characters, and bucket names can contain any
+    # combination of uppercase letters, lowercase letters, numbers, periods
+    # (.), hyphens (-), and underscores (_).
+    if 'Bucket' not in params:
+        return
+    bucket = params['Bucket']
+    if len(bucket) > 255:
+        error_msg = (
+            'Invalid bucket name "%s": Bucket name cannot be '
+            'longer than 255 characters.' % bucket)
+        raise ParamValidationError(report=error_msg)
+    for char in bucket:
```

Comment:
I think it would be acceptable to have a generic error message for any case in which the regex does not match the provided bucket. Right now it's just the difference between too long and invalid character cases. I think it would be fairly obvious that their bucket is too long if you are including the provided bucket in the error message. Even still, you could keep the length check and then use the regex check to have different error messages if desired.